### DISCUSSION TITLE
Cannot read properties of null (reading 'name')

### DISCUSSION THREAD

**Discussion by devajmeireles** [livewire/livewire#8094] [7. Bugs (archived)] [answered]
### Livewire version

Latest

### Laravel version

Latest

### Which PHP version are you using?

PHP 8.1

### Steps To Reproduce

1. Create a component
2. Create a property typed for a Laravel Eloquent Model.
3. Try to bind this property.

```php
<?php

namespace App\Livewire;

use App\Models\User;
use Livewire\Component;

class Test extends Component
{
    public User $user;

    public function render()
    {
        return view('livewire.test');
    }
}
```

```blade
<div>
    <input wire:model="user.name" />
</div>

```

### Repository or Wirebox URL

None, the code examples is enough.

### The console error:

![CleanShot 2024-03-11 at 13 41 52](https://github.com/livewire/livewire/assets/60591772/cc5ab2f9-aa89-4c1f-9af9-1a8590f2bf0c)

### The source of the error:

![CleanShot 2024-03-11 at 13 46 25](https://github.com/livewire/livewire/assets/60591772/22acc759-15fb-456a-983c-cf4c97e67561)

### How to fix:

```php
<?php

namespace App\Livewire;

use App\Models\User;
use Livewire\Component;

class Test extends Component
{
    public User $user;

    function mount(): void
    {
        $this->user = new User();
    }

    public function render()
    {
        return view('livewire.test');
    }
}
```

**I just decided to report this because:**

1. For some strange reason, the same error does not occur when we are dealing with a Livewire form, which also needs `.` to make bind work.
2. This error also affects entangles.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by devajmeireles** [comment]
I was able to confirm that this issue doesn't happen in Livewire 2.

**Comment by samlev** [comment]
What is the route that you're using to access this component? Or if you're not using a full page component, how are you embedding it?

It needs to be able to bind to a user model from somewhere. If you're using it as a full page component, then that has to come from the route via route model binding:

```php
Route('/users/{user}',  Test::class);
```

If you're embedding it into another page, you have to pass the user through:

```html
<livewire:test :user="$user" />
```

If you're using this to create a user (i.e. there is no user to bind to) then what you're doing in mount is correct. Livewire won't new-up an empty eloquent model for you automatically.

I'll also note that unless you have changed the config to enable legacy model binding (which you shouldn't do if this is a new build), then you still won't be able to bind to properties of a model.

The reason why this is showing up as a _javascript_ error is because the uninitialized property `$user` isn't being referenced in the PHP before rendering. If it was, it would be a PHP error.

It doesn't happen with a livewire form object because they have a fixed "shape", so livewire _can_ new up one of those automatically. Eloquent models don't have fixed properties built into the object so creating a new model doesn't actually make the fields from the database available because the model itself doesn't actually know what fields the database has until it pulls a record out.

**Reply by devajmeireles** [reply]
The component is as simple as I demonstrate in the codes above, I didn't even prepare a test repository because I didn't see the need since it is possible to reproduce this easily.

➡️  Regardless of how it is used, it will result in the same errors.

As per my review, I could conclude that the problem is in the `dataGet` function, but I'm not sure if there is something deeper. I enabled `legacy_model_binding` to enable Model binding.

I noticed that when using a Livewire form there are no errors. The most sensible way to solve this is to build the model in `mount`, but what intrigued me is that in Livewire 2 this worked well, that is: I could use `wire:model` for a non-instantiated Model without errors in Livewire 2, something like:

### Livewire 2:

```php
<?php

namespace App\Livewire;

use App\Models\User;
use Livewire\Component;

class Test extends Component
{
    public User $user;

    public function render()
    {
        return view('livewire.test');
    }
}
```

```blade
<div>
    <input wire:model="user.name" />
</div>
```

---

### Livewire 3:

The Livewire 2 code, demonstrated above will result in the console log errors. In the Livewire 3, the "workaround" to avoid the Browser console errors is:

```diff
class Test extends Component
{
    public User $user;

+    function mount(): void
+   {
+       $this->user = new User();
+    }

    public function render()
    {
        return view('livewire.test');
    }
}
```

**Reply by samlev** [reply, marked as answer]
I think that you may have misunderstood. This isn't a bug in Livewire v3 - it's expected behaviour.

It "worked" in Livewire v2 because it just created an empty user model and treated it like an array. You could `wire:model="user.somePropertyThatDoesntExist"` in v2, and you would only find out that it didn't exist when you tried to save the user.

In v3, even if it _did_ create a new user model automatically, you _still_ couldn't `wire:model` or entangle to properties of the user and expect that you could update them. Wiring to Eloquent models and collections is just fraught with peril. See [the official documentation](https://livewire.laravel.com/docs/properties#eloquent-constraints-arent-preserved-between-requests) as well as the discussion on #5586.

This isn't a bug, this is a deliberate change in functionality between v2 and v3.

**Comment by devajmeireles** [comment]
We also have a problem in AlpineJS when we try to 'entangle' a non-instantiated Model property, as mentioned above. When we try to `entangle` this property we get the same error, but now related to AlpineJS:

![CleanShot 2024-03-11 at 23 37 34](https://github.com/livewire/livewire/assets/60591772/36edcbbf-3c62-40db-a6cd-35482f922987)

**Comment by devajmeireles** [comment]
PR with a failure test submitted: https://github.com/livewire/livewire/pull/8098
